packages/flutter/test/cupertino/text_field_restoration_test.dart

```diff
@@ -0,0 +1,126 @@
+  });
+
+  testWidgets('CupertinoTextField restoration with external controller', (WidgetTester tester) async {
+    await tester.pumpWidget(
+      const RootRestorationScope(
+        child: TestWidget(
+          useExternal: true,
+        ),
+        restorationId: 'root',
+      ),
+    );
+
+    await restoreAndVerify(tester);
+  });
+}
+
+Future<void> restoreAndVerify(WidgetTester tester) async {
+  expect(find.text(text), findsNothing);
+  expect(tester.state<ScrollableState>(find.byType(Scrollable)).position.pixels, 0);
+
+  await tester.enterText(find.byType(CupertinoTextField), text);
+  await skipPastScrollingAnimation(tester);
+  expect(tester.state<ScrollableState>(find.byType(Scrollable)).position.pixels, 0);
+
+  await tester.drag(find.byType(Scrollable), const Offset(0, -80));
+  await skipPastScrollingAnimation(tester);
+
+  expect(find.text(text), findsOneWidget);
+  expect(tester.state<ScrollableState>(find.byType(Scrollable)).position.pixels, 60);
+
+  await tester.restartAndRestore();
+
+  expect(find.text(text), findsOneWidget);
+  expect(tester.state<ScrollableState>(find.byType(Scrollable)).position.pixels, 60);
+
+  final TestRestorationData data = await tester.getRestorationData();
+
+  await tester.enterText(find.byType(CupertinoTextField), alternativeText);
+  await skipPastScrollingAnimation(tester);
+  await tester.drag(find.byType(Scrollable), const Offset(0, 80));
+  await skipPastScrollingAnimation(tester);
+
+  expect(find.text(text), findsNothing);
+  expect(tester.state<ScrollableState>(find.byType(Scrollable)).position.pixels, isNot(60));
+
+  await tester.restoreFrom(data);
+
+  expect(find.text(text), findsOneWidget);
+  expect(tester.state<ScrollableState>(find.byType(Scrollable)).position.pixels, 60);
+}
+
+class TestWidget extends StatefulWidget {
+  const TestWidget({Key key, this.useExternal = false}) : super(key: key);
+
+  final bool useExternal;
+
+  @override
+  TestWidgetState createState() => TestWidgetState();
+}
+
+class TestWidgetState extends State<TestWidget> with RestorationMixin {
+  final RestorableTextEditingController controller = RestorableTextEditingController();
+
+  @override
+  String get restorationId => 'widget';
+
+  @override
+  void restoreState(RestorationBucket oldBucket) {
+    registerForRestoration(controller, 'controller');
+  }
+
+  @override
+  void dispose() {
+    controller.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      home: Material(
+        child: Align(
+          alignment: Alignment.center,
+          child: SizedBox(
+            width: 50,
+            child: CupertinoTextField(
+              restorationId: 'text',
+              maxLines: 3,
+              controller: widget.useExternal ? controller.value : null,
+            ),
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+Future<void> skipPastScrollingAnimation(WidgetTester tester) async {
+  await tester.pump();
```

Comment:
Is this pump necessary even with the duration pump on the next line?

Same question for the material text field test.

Reply:
We need one pump to start the animation. And another (with a duration) to finish it.